File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/UserDefinedTableFunctionITCase.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.api.scala.batch

import org.apache.flink.api.scala.batch.utils.TableProgramsTestBase
import org.apache.flink.api.scala.batch.utils.TableProgramsTestBase.TableConfigMode
import org.apache.flink.api.scala._
import org.apache.flink.api.scala.table._
import org.apache.flink.api.table.expressions.utils.{TableFunc0, TableFunc2, TableFunc1}
import org.apache.flink.api.table.{Row, Table, TableEnvironment}
import org.apache.flink.test.util.MultipleProgramsTestBase.TestExecutionMode
import org.apache.flink.test.util.TestBaseUtils
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

import scala.collection.JavaConverters._
import scala.collection.mutable

@RunWith(classOf[Parameterized])
class UserDefinedTableFunctionITCase(
  mode: TestExecutionMode,
  configMode: TableConfigMode)
  extends TableProgramsTestBase(mode, configMode) {

  @Test
  def testUDTF(): Unit = {
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv: BatchTableEnvironment = TableEnvironment.getTableEnvironment(env)
    val in: Table = getSmall3TupleDataSet(env).toTable(tableEnv).as('a, 'b, 'c)
    tableEnv.registerTable("MyTable", in)
    tableEnv.registerFunction("split", TableFunc1)

    val sqlQuery = "SELECT MyTable.c, t.s FROM MyTable CROSS APPLY split(c) AS t(s)"

    val result = tableEnv.sql(sqlQuery).toDataSet[Row]
    val results = result.collect()
    val expected: String = "Jack#22,Jack\n" + "Jack#22,22\n" + "John#19,John\n" + "John#19,19\n" +
      "Anna#44,Anna\n" + "Anna#44,44\n"
    TestBaseUtils.compareResultAsText(results.asJava, expected)

    // with overloading
    val sqlQuery2 = "SELECT MyTable.c, t.s FROM MyTable CROSS APPLY split(c, '$') AS t(s)"
    val result2 = tableEnv.sql(sqlQuery2).toDataSet[Row]
    val results2 = result2.collect()
    val expected2: String = "Jack#22,$Jack\n" + "Jack#22,$22\n" + "John#19,$John\n" +
      "John#19,$19\n" + "Anna#44,$Anna\n" + "Anna#44,$44\n"
    TestBaseUtils.compareResultAsText(results2.asJava, expected2)
  }

  @Test
  def testUDTFCustomReturnType(): Unit = {
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv: BatchTableEnvironment = TableEnvironment.getTableEnvironment(env)
    val in: Table = getSmall3TupleDataSet(env).toTable(tableEnv).as('a, 'b, 'c)
    tableEnv.registerTable("MyTable", in)
    tableEnv.registerFunction("split", TableFunc2)

    val sqlQuery = "SELECT MyTable.c, t.a, t.b  FROM MyTable CROSS APPLY split(c) AS t(a,b)"

    val result = tableEnv.sql(sqlQuery).toDataSet[Row]
    val results = result.collect()
    val expected: String = "Jack#22,Jack,4\n" + "Jack#22,22,2\n" + "John#19,John,4\n" +
      "John#19,19,2\n" + "Anna#44,Anna,4\n" + "Anna#44,44,2\n"
    TestBaseUtils.compareResultAsText(results.asJava, expected)
  }

  @Test
  def testUDTFWithOuterApply(): Unit = {
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv: BatchTableEnvironment = TableEnvironment.getTableEnvironment(env)
    val in: Table = getSmall3TupleDataSet(env).toTable(tableEnv).as('a, 'b, 'c)
    tableEnv.registerTable("MyTable", in)
    tableEnv.registerFunction("split", TableFunc2)

    val sqlQuery = "SELECT MyTable.c, t.a, t.b  FROM MyTable OUTER APPLY split(c) AS t(a,b)"

    val result = tableEnv.sql(sqlQuery).toDataSet[Row]
    val results = result.collect()
    val expected: String = "Jack#22,Jack,4\n" + "Jack#22,22,2\n" + "John#19,John,4\n" +
      "John#19,19,2\n" + "Anna#44,Anna,4\n" + "Anna#44,44,2\n" + "nosharp,null,null"
    TestBaseUtils.compareResultAsText(results.asJava, expected)
  }

  @Test
  def testUDTFWithFilter(): Unit = {
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv: BatchTableEnvironment = TableEnvironment.getTableEnvironment(env)
    val in: Table = getSmall3TupleDataSet(env).toTable(tableEnv).as('a, 'b, 'c)
    tableEnv.registerTable("MyTable", in)
    tableEnv.registerFunction("split", TableFunc0)

    val sqlQuery = "SELECT MyTable.c, t.name, t.age " +
      "FROM MyTable CROSS APPLY split(c) AS t(name,age) " +
      "WHERE t.age > 20"

    val result = tableEnv.sql(sqlQuery).toDataSet[Row]
    val results = result.collect()
    val expected: String = "Jack#22,Jack,22\n" + "Anna#44,Anna,44\n"
    TestBaseUtils.compareResultAsText(results.asJava, expected)
  }

  @Test
  def testUDTFWithScalarTableAPI(): Unit = {
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv: BatchTableEnvironment = TableEnvironment.getTableEnvironment(env)
    val in: Table = getSmall3TupleDataSet(env).toTable(tableEnv).as('a, 'b, 'c)

    val table = in.crossApply(TableFunc0('c) as ('name, 'age))
      .select('c, 'name, 'age)
      .where('age > 20)

    val result = table.toDataSet[Row]
    val results = result.collect()
    val expected: String = "Jack#22,Jack,22\n" + "Anna#44,Anna,44\n"
    TestBaseUtils.compareResultAsText(results.asJava, expected)
  }

  @Test
  def testUDTFWithScalarTableAPIAndOuterApply(): Unit = {
    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
    val tableEnv: BatchTableEnvironment = TableEnvironment.getTableEnvironment(env)
    val in: Table = getSmall3TupleDataSet(env).toTable(tableEnv).as('a, 'b, 'c)

    val table = in.outerApply(TableFunc2('c) as ('name, 'age))
      .select('c, 'name, 'age)

    val result = table.toDataSet[Row]
    val results = result.collect()
    val expected: String = "Jack#22,Jack,4\n" + "Jack#22,22,2\n" + "John#19,John,4\n" +
      "John#19,19,2\n" + "Anna#44,Anna,4\n" + "Anna#44,44,2\n" + "nosharp,null,null"
    TestBaseUtils.compareResultAsText(results.asJava, expected)
  }

  private def getSmall3TupleDataSet(env: ExecutionEnvironment): DataSet[(Int, Long, String)] = {
    val data = new mutable.MutableList[(Int, Long, String)]
    data.+=((1, 1L, "Jack#22"))
    data.+=((2, 2L, "John#19"))
    data.+=((3, 2L, "Anna#44"))
    data.+=((4, 3L, "nosharp"))
    env.fromCollection(data)
  }
}

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/UserDefinedTableFunctionITCase.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.api.java.batch;

import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.table.BatchTableEnvironment;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.tuple.Tuple5;
import org.apache.flink.api.scala.batch.utils.TableProgramsTestBase;
import org.apache.flink.api.table.Row;
import org.apache.flink.api.table.Table;
import org.apache.flink.api.table.TableEnvironment;
import org.apache.flink.api.table.functions.TableFunction;
import org.apache.flink.test.javaApiOperators.util.CollectionDataSets;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.List;


@RunWith(Parameterized.class)
public class UserDefinedTableFunctionITCase extends TableProgramsTestBase {

	public UserDefinedTableFunctionITCase(TestExecutionMode mode, TableConfigMode configMode){
		super(mode, configMode);
	}


	@Test
	public void testUDTF() throws Exception {
		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
		BatchTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);

		DataSet<Tuple5<Integer, Long, Integer, String, Long>> ds1 =
			CollectionDataSets.getSmall5TupleDataSet(env);

		Table table = tableEnv.fromDataSet(ds1, "a, b, c, d, e");

		tableEnv.registerFunction("stack", new TableFunc0());

		Table result = table.crossApply("stack(a,c) as (f)")
			.select("b,f");

		// with overloading
		DataSet<Row> ds = tableEnv.toDataSet(result, Row.class);
		List<Row> results = ds.collect();
		String expected = "1,1\n" + "1,0\n" + "2,2\n" + "2,1\n" + "3,2\n" + "3,2\n";
		compareResultAsText(results, expected);

		Table result2 = table.crossApply("stack(a,c,e) as (f)")
			.select("b,f");

		DataSet<Row> ds2 = tableEnv.toDataSet(result2, Row.class);
		List<Row> results2 = ds2.collect();
		String expected2 = "1,1\n" + "1,1\n" + "1,0\n" + "2,2\n" + "2,2\n" + "2,1\n" +
		                   "3,1\n" + "3,2\n" + "3,2\n";
		compareResultAsText(results2, expected2);
	}

	@Test
	public void testUDTFWithOuterApply() throws Exception {
		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
		BatchTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);

		DataSet<Tuple5<Integer, Long, Integer, String, Long>> ds1 =
			CollectionDataSets.getSmall5TupleDataSet(env);

		Table table = tableEnv.fromDataSet(ds1, "a, b, c, d, e");

		tableEnv.registerFunction("func1", new TableFunc1());

		Table result = table.crossApply("func1(d) as (s,l)")
			.select("d,s,l");

		DataSet<Row> ds = tableEnv.toDataSet(result, Row.class);
		List<Row> results = ds.collect();
		String expected = "Hallo Welt,Welt,4\n" + "Hallo Welt wie,Welt,4\n" +
		                  "Hallo Welt wie,wie,3\n";
		compareResultAsText(results, expected);


		Table result2 = table.outerApply("func1(d) as (s,l)")
			.select("d,s,l");

		DataSet<Row> ds2 = tableEnv.toDataSet(result2, Row.class);
		List<Row> results2 = ds2.collect();
		String expected2 = "Hallo,null,null\n" + "Hallo Welt,Welt,4\n" + "Hallo Welt wie,Welt,4\n" +
		                  "Hallo Welt wie,wie,3\n";
		compareResultAsText(results2, expected2);
	}

	@Test
	public void testUDTFWithScalarFunction() throws Exception {
		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
		BatchTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);

		DataSet<Tuple5<Integer, Long, Integer, String, Long>> ds1 =
			CollectionDataSets.getSmall5TupleDataSet(env);

		Table table = tableEnv.fromDataSet(ds1, "a, b, c, d, e");

		tableEnv.registerFunction("func0", new TableFunc0());

		Table result = table.crossApply("func0(c, charLength(d)) as (l)")
			.select("d,l");

		DataSet<Row> ds = tableEnv.toDataSet(result, Row.class);
		List<Row> results = ds.collect();
		String expected = "Hallo,0\n" + "Hallo,5\n" + "Hallo Welt,1\n" + "Hallo Welt,10\n" +
		                  "Hallo Welt wie,2\n" + "Hallo Welt wie,14\n";
		compareResultAsText(results, expected);
	}


	public static class TableFunc0 extends TableFunction<Integer> {
		public void eval(int a, int b) {
			collect(a);
			collect(b);
		}

		public void eval(int a, int b, long c) {
			collect(a);
			collect(b);
			collect((int) c);
		}
	}

	public static class TableFunc1 extends TableFunction<Tuple2<String, Integer>> {
		public void eval(String str) {
			for (String s : str.split(" ")) {
				if (!s.equals("Hallo")) {
					collect(new Tuple2<>(s, s.length()));
				}
			}
		}
	}

}
